### DISCUSSION TITLE
Side Effects of Setting Handle.id Directly?

### DISCUSSION THREAD

**Discussion by zicklag** [bevyengine/bevy#6171] [Q&A]
If I understand correctly, because the `id` field of `Handle<T>` is public, if I set that field directly on a strong handle, it would fail to decrement the strong reference count for the previous handle ID, and then the asset previously referenced by the strong handle would never be dropped.

Am I understanding that correctly?

**Comment by mockersf** [comment]
I'm not sure on all the side effects that would have, but what you're describing seems correct 👍 

The `id` field is public more for read access than write access... I guess a side effect of you opening that discussion will be a PR to make it private with an accessor 😄

**Reply by zicklag** [reply]
Before you make it private, I'm actually needing the ability to set that! Though it's current state definitely isn't right, it'd be good if we could find a way to fix the issue while still allowing what I need.

My use-case is that I need to be able to somehow assign to a `Handle<T>` through `Box<dyn Reflect>` when all I have is the `Box<dyn Reflect>` of a handle ID or an `UntypedHandle` ( though `UntypedHandle`s don't implement `Reflect` currently ).

In other words, I need this to work:

```rust
use bevy::{
    asset::{AssetPath, HandleId},
    prelude::*,
    reflect::GetPath,
};

#[derive(Reflect, Default, Debug, Component)]
struct MyComponent {
    handle: Handle<TextureAtlas>,
}

fn main() {
    let component = MyComponent::default();
    let mut reflected_component = Box::new(component) as Box<dyn Reflect>;

    let asset_path: AssetPath = "test".into();
    let handle_id: HandleId = asset_path.into();

    let reflected_handle_id = Box::new(handle_id) as Box<dyn Reflect>;

    reflected_component
        .path_mut("handle.id")
        .unwrap()
        .set(reflected_handle_id)
        .unwrap();
}
```

I'm needing this so that I can set things like sprite image handles in my scripting implementation.

**Reply by mockersf** [reply]
The "proper" way to go from a `HandleId` to a (strong) `Handle<T>` is through [`AssetServer::get_handle`](https://docs.rs/bevy/0.8.1/bevy/asset/struct.AssetServer.html#method.get_handle) or [`Assets<T>::get_handle`](https://docs.rs/bevy/0.8.1/bevy/asset/struct.Assets.html#method.get_handle). Would those work for you?

I don't think there's a way currently to go from a `HandleId` to a weak `Handle<T>`, but I'm not sure that would be useful...

**Reply by zicklag** [reply]
I could use `AssetServer::get_handle_untyped` to get an `UntypedHandle` if `UntypedHandle` implemented `Reflect`. I don't know the type of the asset I'm loading when the script runs, so I can't use a typed handle in the script.

Then there is still the issue that I can't assign an untyped handle to a `Handle<T>` if that `Handle<T>` is currently in a `&mut dyn Reflect`.

Setting the ID worked because I can get the `id` field through the Reflect trait and then assign to it.

:thinking: 

If we had reflect functions, like is being worked on [here](https://github.com/jakobhellermann/bevy_reflect_fns) then we could expose a reflect function to create a strong handle from an untyped handle or a weak handle from a handle ID. That's probably the correct solution.

So maybe we do just make this private in Bevy and I'll just need to try and leverage reflect functions to do the conversion. Or I might just find another workaround for my use-case, 'cause I'm still working on fleshing it out.

**Reply by mockersf** [reply]
I hate what I'm going to write, but... making the field private doesn't stop you from changing it using reflection.

You'll just be committing two crimes instead of one 😄. And it would cleanup the public API

**Reply by zicklag** [reply]
LOL. I guess I might just be a criminal, then.

**Reply by mockersf** [reply]
#6176

**Comment by zicklag** [comment]
Looks like all of my concerns as far as scripting are covered by this PR: https://github.com/bevyengine/bevy/pull/5923! 🎉
